src/merkle/tiered_smt/tests.rs

```diff
@@ -0,0 +1,216 @@
+
+        // append the leaves to the engine.
+        leaves.iter().for_each(|leaf| {
+            test.insert(leaf.key, leaf.value);
+        });
+
+        // assert the values are available. the tree will be able to provide the value if, and only
+        // if, it can traverse from the root to the leaf - therefore, if they are available, the
+        // full path is also correct.
+        leaves.iter().for_each(|leaf| {
+            let value = test.get_leaf_value(&leaf.key.into()).unwrap().unwrap();
+            assert_eq!(value, leaf.value);
+        });
+    }
+}
+
+// TIERED SPARSE MERKLE TREE TEST ENGINE
+// ================================================================================================
+
+pub struct TieredTestEngine {
+    tree: TieredSmt,
+}
+
+impl Default for TieredTestEngine {
+    fn default() -> Self {
+        let storage = Storage::default();
+        let tree = TieredSmt::with_storage(storage).unwrap();
+        Self { tree }
+    }
+}
+
+impl Deref for TieredTestEngine {
+    type Target = TieredSmt;
+
+    fn deref(&self) -> &Self::Target {
+        &self.tree
+    }
+}
+
+impl DerefMut for TieredTestEngine {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.tree
+    }
+}
+
+impl TieredTestEngine {
+    pub fn insert<K, V>(&mut self, key: K, value: V) -> &mut Self
+    where
+        Word: From<K>,
+        Word: From<V>,
+    {
+        let key = Word::from(key);
+        let value = Word::from(value);
+
+        // assert non-membership
+        let root = self.tree.root();
+        let proof = self.tree.prove_non_membership(key, value).unwrap();
+        assert!(proof.verify(root));
+
+        self.tree.insert(key, value).unwrap();
+
+        // assert membership
+        let root = self.tree.root();
+        let proof = self.tree.prove_membership(key).unwrap().unwrap();
+        assert!(proof.verify(root));
+
+        self
+    }
+
+    pub fn expect_path<I, W, WI>(&mut self, key: W, path: I) -> &mut Self
+    where
+        Word: From<W>,
+        Word: From<WI>,
+        I: IntoIterator<Item = WI>,
+    {
+        let key = Word::from(key);
+        let opening = self.tree.get_path(&key).unwrap().unwrap();
+        assert_eq!(
+            opening.deref(),
+            path.into_iter().map(Word::from).collect::<Vec<_>>()
+        );
+        self
+    }
+
+    /// create an iterator that assumes the first bit of the tier is flipped, relative to a mutated
+    /// sibling.
+    pub fn tier_iterator(&mut self, key: &Word, depth: u8) -> impl Iterator<Item = RpoDigest> {
+        let mut index = LeafIndex::from(key);
+        index.traverse_to(depth);
+        let mut index = NodeIndex::from(index);
+        (0..15).for_each(|_| index.reverse());
+        EMPTY_SUBTREES[depth as usize - 14..depth as usize + 1]
+            .iter()
+            .rev()
+            .copied()
+            .chain([self.get_node(&index.sibling()).unwrap().into()])
+    }
```

Comment:
I usually avoid using the structure I'm testing to craft the expected output of the tests as it might create false positives.

One alternative to that would be having an oversimplified storage inside the test engine, and clash both implementations to assert correctness. I would prefer that over using the original tree to test itself. Wdyt?

cc @Al-Kindi-0 ; to keep the discussion in one place, better we continue here.

Reply:
I think we could have storage-specific tests in storage, and then here we can assume that the storage works correctly. Otherwise, we'd pretty much be duplicating the implementation of storage here, no?

Reply:
@vlopes11 I am thinking more of something like going through an algorithm using pen and paper using a small example. It is true that we are in essence doing the same thing but the more primitive approach has other benefits e.g. testing the logic or documenting it.
The pen and paper in our case will be the potentially duplicated code but simplified through the specificity of the example we are going through.
As @bobbinth mentioned above, the duplication shouldn't be everywhere as we can assume some parts to be correct e.g. storage.